### PR TITLE
feat(skill): darwinian-evolver — evolutionary optimizer for prompts, regex, SQL, and code

### PR DESCRIPTION
**Pull request by Bihruze**
## Summary

Adds a new optional skill — **`darwinian-evolver`** — that evolves text artifacts (prompts, regexes, SQL queries, small code snippets) toward a user-supplied fitness function via LLM-driven mutation and crossover over a quality-diversity archive.

References issues:
- **#336** — Darwinian Evolver Skill (this is the scoped v1 the issue asks for)
- **#337** — Evolutionary Self-Improvement (this PR lands Phase 3 in skill form)
- Bridges to `NousResearch/hermes-agent-self-evolution` via a Tier 3 DSPy-jsonl export — no import, just a data contract.

No other skill or tool in the repo performs prompt / code optimization today (verified by scanning `skills/` and `optional-skills/` for DSPy / GEPA / evolutionary keywords).

## Architecture — three tiers

```
┌──────────────────────────────────────────────────────────────────────┐
│ Tier 3: bridge                                                        │
│   export lineage → DSPy-compatible JSONL / GEPA reflective trace      │
│                 → hermes-agent-self-evolution ingests                 │
├──────────────────────────────────────────────────────────────────────┤
│ Tier 2: heavy (opt-in, AGPL-isolated)                                 │
│   subprocess wrapper around openevolve CLI (Apache 2.0)               │
│   subprocess wrapper around darwinian-evolver CLI (Imbue, AGPL v3)    │
│   NO python import — mere aggregation — license-clean                 │
├──────────────────────────────────────────────────────────────────────┤
│ Tier 1: core (MIT, default)                                           │
│   (μ+λ)-ES · MAP-Elites · NSGA-II · tournament + rank selection       │
│   LLM operators: paraphrase · structural-edit · semantic-crossover    │
│                   · CoT-inject · novelty-seeking · critique-then-edit │
│                   · PromptBreeder-style meta-mutator                  │
│   Evaluator: async batch · successive halving · held-out guard        │
│   Storage: SQLite lineage · content-addressed genomes · Mermaid DAGs  │
└──────────────────────────────────────────────────────────────────────┘
              reuses Hermes's existing OpenAI-compat client
              (prompt caching aware; respects per-slot context from #12595)
```

Tier 1 has no non-stdlib runtime deps beyond `httpx` (already in Hermes core). Tiers 2 and 3 are thin adapters that fail gracefully when their externals are absent.

## Algorithmic core (with citations)

| Primitive | Reference |
|---|---|
| (μ+λ)-ES survival | Bäck & Schwefel 1993 |
| MAP-Elites quality-diversity archive | Mouret & Clune 2015 |
| NSGA-II fast non-dominated sort + crowding distance | Deb et al. 2002 |
| LLM semantic crossover | Meyerson et al. 2023 |
| Self-referential meta-mutator | Fernando et al. 2023 (PromptBreeder) |
| Reflective critique-then-edit | Agrawal et al. 2025 (GEPA) |
| Novelty-seeking mutation bias | Lehman & Stanley 2011 |
| Exp3 bandit over operators | Auer et al. 2002 |
| Successive halving (Hyperband-lite) | Li et al. 2017 |

All primitives are implemented from scratch (no large dependencies) and exposed as pure-function library code in `scripts/algorithms.py`; `evolver.py` composes them.

## Fitness contract

Users drop a `fitness.py` into their experiment directory:

```python
from evolver_sdk import fitness_spec

@fitness_spec(held_out_frac=0.2, timeout_s=30, objectives=[\"accuracy\", \"cost\"])
def fitness(candidate: str, context: dict) -> float | dict[str, float]:
    ...
```

The evaluator guarantees:

- Seed propagation (`context[\"seed\"]`) for reproducible LLM calls (OpenAI / Anthropic / vLLM all honor `seed` as a hint or hard param).
- Code candidates execute in a subprocess sandbox with `resource.setrlimit` caps (CPU, address space, data segment) plus a wall-clock timeout.
- A **reward-hacking guard** re-scores the top-K on a held-out split each generation; candidates with a >15 % generalization gap take a penalty so honest candidates outrank overfitters.
- A global **`--budget`** (USD or tokens via rate args) hard-kills runs via `BudgetExceeded` before blowing past the cap.

## License handling

Imbue's `darwinian-evolver` is **AGPL v3**. This PR never imports it. The `ExternalEvolverAdapter` wraps it as an opaque subprocess invocation (\"mere aggregation\" exemption) and gates on `shutil.which`, raising a clear `AdapterUnavailable` with an install hint when absent. The install hint surfaces the license explicitly (\"AGPL v3 — review license before use\"). The default Tier 2 backend is OpenEvolve (Apache 2.0) so users never touch AGPL unless they explicitly opt in.

## Validation

`tests/skills/test_darwinian_evolver.py` — **39 cases, all green** locally:

| area | cases | covers |
|---|---:|---|
| storage | 6 | content-address determinism, idempotent insert, ancestry reconstruction, held-out-preferring best lookup, budget totals, lineage_hash stability + change detection |
| algorithms | 9 | tournament bias (statistical), rank-select invariant, (μ+λ) elitism, MAP-Elites place/coverage/empty-sample, NSGA-II front identification, strict dominance, crowding-distance boundary infinity, Exp3 reward clamp + weight update |
| evaluator | 4 | fitness_spec metadata round-trip, scalar + dict fitness via async batch, timeout → worst-score fallback |
| reward hacking | 1 | overfitter penalised >0.5 vs honest 0.0 penalty |
| successive halving | 1 | 8 → 4 → 2 narrowing at the expected fidelity schedule |
| sandbox | 4 | simple program, syntax error, runaway `while True` killed within wall-clock, pytest summary parser |
| adapters | 5 | openevolve + Imbue graceful absence with license hint, DSPy-jsonl default-best-per-gen, DSPy-jsonl --all, GEPA trace filters to reflective edges |
| LLM client | 2 | `seed` propagated into request body, BudgetLedger raises at cap across multi-call runs |
| MAP-Elites coverage | 1 | random 200 samples fill ≥60 % of 4×4 grid |
| end-to-end (mocked LLM) | 1 | single generation improves over seed + stable lineage hash |

### Acceptance checklist (10/10)

1. Determinism — ✅ `lineage_hash` stability asserted.
2. Budget hard-kill — ✅ `BudgetExceeded` verified.
3. Multi-objective correctness — ✅ NSGA-II front non-domination asserted.
4. MAP-Elites coverage — ✅ ≥60 % on random 200-sample run.
5. Sandbox kills runaway — ✅ `while True` killed within wall-clock + overhead.
6. Held-out guard — ✅ overfitter penalty ≥0.5.
7. End-to-end improves — ✅ mocked-LLM generation improves fitness over seed.
8. Determinism of LLM ops — ✅ `seed` in request body asserted.
9. Tier 2 isolation — ✅ both adapters raise `AdapterUnavailable` with install hint.
10. CI regression — full `tests/skills/test_darwinian_evolver.py` green.

## Scope — explicit guardrails

Shipped:

- Tier 1 core, 7 LLM operators, MAP-Elites + ES + NSGA-II, held-out + budget guards, SQLite lineage, Mermaid export, replay.
- Tier 2 adapters (`openevolve`, `darwinian-evolver`).
- Tier 3 DSPy + GEPA exports.
- 5 fitness templates (prompt / regex / sql / code / multi-objective).
- End-to-end `summarize_10_words` demo with deterministic fitness.

**Not** in scope for v0.1 (deliberately deferred):

- RL fine-tuning (Phase 4 of #337).
- UI dashboard; Mermaid + JSON only.
- Distributed / multi-node execution.
- Automatic hyperparameter search.
- Importing Imbue's AGPL Python — CLI only, always.

## File map

```
optional-skills/research/darwinian-evolver/
├── SKILL.md                        (full author spec + theoretical foundations)
├── scripts/
│   ├── evolver.py                  (CLI: init/run/status/best/lineage/budget/export/replay)
│   ├── algorithms.py               (ES, MAP-Elites, NSGA-II, Exp3, descriptors)
│   ├── operators.py                (7 LLM operators + segment-splice + meta-mutator)
│   ├── evaluator.py                (fitness_spec, async batch, held-out guard, halving)
│   ├── sandbox.py                  (subprocess + rlimit + timeout for code eval)
│   ├── storage.py                  (SQLite lineage, content-addressed IDs, hashing)
│   ├── llm.py                      (async OpenAI-compat client + BudgetLedger)
│   └── adapters.py                 (Tier 2 CLI wrappers + Tier 3 DSPy/GEPA export)
├── templates/                      (5 copy-paste fitness templates)
└── demos/summarize_10_words/       (end-to-end demo with deterministic fitness)
tests/skills/test_darwinian_evolver.py   (39 cases)
```

## Diff shape

~2,900 insertions across 17 new files + minor edits to the two files that existed in my first commit. Total skill surface is ~2,000 LOC production, ~600 LOC tests, and ~300 LOC templates/demos.

## Note on the directory rename

The repo-level `.gitignore` includes `examples/`. To keep the end-to-end walkthrough tracked without touching the ignore list, I placed it under `demos/summarize_10_words/`. Happy to change the name if maintainers prefer a different convention.

### CODE REVIEW

**Comment by Bihruze** [comment]
## v0.2 changelog (pushed as `1deec86`)

Layered on top of the two v0.1 commits — four cohesive features that move the skill from credible-weekend to production-grade. All opt-in except the cache, which is transparent.

### Summary

| # | Feature | Why it matters | New LOC |
|---|---|---|---|
| 1 | **LLM response cache** — content-addressed SQLite, keyed by blake2b-16 of the normalised request body | Replay is now bit-for-bit deterministic at zero cost. Cache hits skip both the HTTP call and the budget ledger. | `cache.py` (150) + edits |
| 2 | **Bradley-Terry pairwise judge** — Hunter (2004) MM aggregator over position-bias-guarded pairwise verdicts | Absolute LLM-as-judge scores drift (Zheng 2023, Chen 2024). Pairwise preferences are invariant to that drift. | `judge.py` (250) + evaluator wiring |
| 3 | **Constitutional critic** — second LLM inspects top-K for reward hacking per a markdown rule-set | Held-out guard catches train/test gaps; the critic catches *generalised* cheating (judge flattery, regex over-matching, harness exploits). | `critic.py` (180) + `constitution.md` |
| 4 | **FastAPI + Plotly dashboard** — read-only live view with Mermaid lineage | Runs become visible — fitness curves, Pareto, operator attribution, budget burndown — with WebSocket push on new generations. | `dashboard.py` (230) + `dashboard.html` (180) |

### CLI additions

```
evolver run <dir> [--no-cache]
evolver cache <dir> stats|purge
evolver dashboard <dir> [--host 127.0.0.1 --port 8787]
```

### `fitness_spec` extensions

```python
@fitness_spec(
    judge="pairwise", pairwise_rounds=40,        # feature 2
    critic="on", critic_threshold=0.5,           # feature 3
    critic_top_k=5, critic_model="haiku-cheap",  # feature 3 — cheap-model route
)
def fitness(candidate, context): ...
```

### Storage

Four additive tables (`CREATE TABLE IF NOT EXISTS` — v0.1 experiments auto-upgrade): `llm_cache`, `pairwise_votes`, `bt_scores`, `critic_evaluations`.

### Tests

31 new cases across four test files, all green:

```
test_darwinian_evolver_v02_cache.py       8 cases
test_darwinian_evolver_v02_judge.py      12 cases
test_darwinian_evolver_v02_critic.py      6 cases
test_darwinian_evolver_v02_dashboard.py   5 cases
```

Full darwinian-evolver suite: **39 (v0.1) + 31 (v0.2) = 70/70 green**. No regression in other skills.

Acceptance rows 11 (replay bit-identity) and 12 (judge calibration via Condorcet recovery) from the v0.2 plan both pass.

### License / safety

* Dashboard binds `127.0.0.1` by default; `--host` non-loopback prints a prominent warning. No auth shipped.
* FastAPI / uvicorn are optional deps; dashboard subcommand fails closed with an install hint when absent.
* Constitutional critic is off by default.
* Cache is per-experiment (scoped to its `lineage.db`), so cross-experiment leakage is structurally impossible.

### What's explicitly out of scope for v0.2

No RL fine-tuning (that's Phase 4 of #337); no dashboard auth/write endpoints; no distributed workers; no automatic hyperparameter tuning.

Happy to split the commit into four (one per feature) if that eases review.

**Comment by Bihruze** [comment]
Closing while v1.0 (A1-C5) lands locally; will reopen with the full roadmap once Bihruze reviews. See /Users/seher/.claude/plans/nested-gathering-kazoo.md for the plan.
